### ISSUE TITLE
[Pull To Refresh][Bug] Glitchy behavior when using preventDefault() on a touch event

### ISSUE THREAD

**Issue by mstange** [mozilla-mobile/fenix#16575]
## Steps to reproduce

 1. Load one of these testcases: [scrollable testcase](https://bugzilla.mozilla.org/attachment.cgi?id=9187857) [non-scrollable testcase](https://bugzilla.mozilla.org/attachment.cgi?id=9187858)
 2. Do a downwards pan gesture on the box with the black border. (no pull to refresh happens so far, good.)
 3. To a short upwards pan gesture outside the box.

### Expected behavior

The short upwards pan should scroll down (on the scrollable variant) or do nothing (on the non-scrollable variant).

### Actual behavior

During the second pan, the pull-to-refresh arrow appears out of nowhere, and a pull-to-refresh is triggered even though the second pan was in the opposite direction.

### Device information

* Android device: Moto G5
* Fenix version: Nightly 201112

**Comment by hiikezoe** [comment]
I've debugged this case, what I've noticed are;

1) in the first downwards pan, GeckoView returns INPUT_RESULT_HANDLED_CONTENT, it looks correct to me
2) in the second upwards pan, GeckoView returns INPUT_RESULT_HANDLED, it also looks correct to me
3) [`(child.getInputResult() == EngineView.InputResult.INPUT_RESULT_HANDLED_CONTENT)`](https://searchfox.org/mozilla-mobile/rev/bafe96753c2c15f3a8bcdbc8842378351e48e786/android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt#78) in canChildScrollUp looks clearly incorrect to me, in HANDLED_CONTENT cases the browser should do nothing since the event has been consumed in the content
4) changing the condition to `(child.getInputResult() == EngineView.InputResult.INPUT_RESULT_HANDLED)` seems to fix the case
5) but still [`child.canScrollVerticallyUp() ||`](https://searchfox.org/mozilla-mobile/rev/bafe96753c2c15f3a8bcdbc8842378351e48e786/android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt#77) looks incorrect to me, given that if a div box having preventDefault() is positioned in the middle of the large root scrollable element, any touchstart events in the div element should trigger scrolling

**Comment by hiikezoe** [comment]
@Mugurell ^ FYI

**Comment by hiikezoe** [comment]
>     4. changing the condition to `(child.getInputResult() == EngineView.InputResult.INPUT_RESULT_HANDLED)` seems to fix the case

A test case in [bug 1677413](https://bugzilla.mozilla.org/show_bug.cgi?id=1677413) makes me notice that this change is not sufficient, actually it's wrong.

Now I suppose GeckoView needs to expose info that we are now at the top edge in all relevant scrollable elements and the upwards pan is not consumed in any contents thus it's time to trigger "pull-to-refresh"?

I don't yet understand when "pull-to-refresh" should happen, though.

CC @theres-waldo

**Comment by Mugurell** [comment]
> 3. [`(child.getInputResult() == EngineView.InputResult.INPUT_RESULT_HANDLED_CONTENT)`](https://searchfox.org/mozilla-mobile/rev/bafe96753c2c15f3a8bcdbc8842378351e48e786/android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt#78) in canChildScrollUp looks clearly incorrect to me, in HANDLED_CONTENT cases the browser should do nothing since the event has been consumed in the content

Based on this condition pull to refresh will not show up.
The page could be scrolled all the way to the top but if the website handled the motion `canChildScrollUp()` will return true and so the pull to refresh throbber will not appear.

>  5. but still [`child.canScrollVerticallyUp() ||`](https://searchfox.org/mozilla-mobile/rev/bafe96753c2c15f3a8bcdbc8842378351e48e786/android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SwipeRefreshFeature.kt#77) looks incorrect to me, given that if a div box having preventDefault() is positioned in the middle of the large root scrollable element, any touchstart events in the div element should NOT trigger scrolling

I think this relates to GV and the app having their own separate touch listeners.
(Would be really awesome for the app to not have to)

Haven't tested but I guess what happens here is that:
- the gesture starts in a div based on which GV returns `HANDLED_CONTENT`
  -  based on the check from `canScrollVerticallyUp()` we don't do nothing.
But we still know about the gesture. We just don't act upon it.
- the finger gets past the bounds of that div, GV now returns `HANDLED` or `UNHANDLED`
  - based on the check from `canScrollVerticallyUp()` we now immediately see a scroll for the entire distance.

If the above reasoning is correct maybe we could somehow "forget" about the distance scrolled if GV returned `HANDLED_CONTENT`..
Or GV could offer an API similar to [canScrollVerticallyUp](https://github.com/mozilla-mobile/android-components/blob/5a3b15e58d32ff9d5a84551010a0bac9a91748b4/components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineView.kt#L143) but better. 
I think GV truly knows if a webpage can be scrolled or not - as opposed to clients trying to figure it out.
Exposing this data would also help us greatly with the dynamic toolbars.
Would be really nice to have complete APIs for looking into what happens in GV - like horizontal / vertical scrolls, flings, etc.

**Comment by theres-waldo** [comment]
> I think GV truly knows if a webpage can be scrolled or not - as opposed to clients trying to figure it out.
> Exposing this data would also help us greatly with the dynamic toolbars.
> Would be really nice to have complete APIs for looking into what happens in GV - like horizontal / vertical scrolls, flings, etc.

Makes sense to me.

cc @snorp

If new GV APIs require APZ to expose additional information, I'm happy to consider that.

A simple formulation or starting point could be to make the `HANDLED_` flags richer, or make them bit-flags or something like that.

**Comment by hiikezoe** [comment]
> or make them bit-flags or something like that.

Yes, that's what I was supposing. Given that the current scroll position in a content is top and the content height is 100vh, thus we can do either "pull-to-refresh" or "hide dynamic toolbar" depending on subsequent touch move directions.  But if the content height is 100%, then the "hide dynamic toolbar" shouldn't happen, we need to differentiate the cases.

**Comment by hiikezoe** [comment]
I tried a native implementation that GeckoView returns a new INPUT_RESULT flag when the content's scroll position is at top and the initial touch start event is handled in the root APZC, but it didn't solve this issue at all.

Then I noticed that at the beginning of the second __upwards__ pan, GeckoView doesn't know which direction user is going to pan. So triggering "pull-to-refresh" is quite weird on the upwards pan regardless of what result GeckoView returns.

Then I changed [fragment_browser.xml](https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/res/layout/fragment_browser.xml) to use SwipeRefreshLayout, then I can no longer see this issue. So I am now supposing this is a regression by using VerticalSwipeRefreshLayout.
